### PR TITLE
limit lock count for scan lock request

### PR DESCRIPTION
**Pull request by wentaoxu**
limit the lock num of one scan-lock request.
If there are too many locks in one region, for example, larger than 400MB, scan-lock request will last very long time, 12min in my case, it would case the gc job run overtime.
This problem happened some times, and the gc worker could not resolve an single lock or clean one overtime row version in this case.
PTAL @disksing @zhangjinpeng1987 
shouldn't be merged before [this pr in kvproto](https://github.com/pingcap/kvproto/pull/220)

### CODE REVIEW

**Comment by wentaoxu** [comment]
PTAL @disksing @zhangjinpeng1987 @siddontang

**Comment by disksing** [comment]
I think we need to support limit in mock-tikv, and we also need a test case that a region has more than limit locks.

**Comment by wentaoxu** [comment]
@disksing, yes, I will do it.

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -467,11 +469,18 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
...
-		key = loc.EndKey
-		if len(key) == 0 {
-			break
+		if len(locks) < gcScanLockLimit {
+			regions++
+			key = loc.EndKey
+			if len(key) == 0 {
+				break
+			}
+			req.ScanLock.StartKey = []byte("")
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Use region start key?

**Comment by wentaoxu** [comment]
@disksing It is a bit difficult to add ut for ScanLock, because we can't call `prewrite` to product locks. Maybe we need to do refactor and pull functions like `resolveLocks` out of the package `gc_worker`, into tikv.
I think it is worthy, but I am prefer to do this in next pr.

File: store/tikv/gcworker/gc_worker.go (hunk 1 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: I think it is a good idea.

**Comment by wentaoxu** [comment]
PTAL @zhangjinpeng1987 @disksing, keep 'req.ScanLock.StartKey' the same as 'key' to avoid the error 'key is not in region' when the region split for several pieces, or merge into one piece.

File: store/tikv/gcworker/gc_worker.go [hunk 2]
```diff
@@ -467,11 +470,20 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
 			}
 			continue
 		}
-		regions++
+
 		totalResolvedLocks += len(locks)
-		key = loc.EndKey
-		if len(key) == 0 {
-			break
+		// we'd better keep 'req.ScanLock.StartKey' the same as 'key' to avoid
```
Comment (disksing) [inline store/tikv/gcworker/gc_worker.go]: stale note?

Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: yes, I will move it to the right place

File: Gopkg.lock [hunk 3]
```diff
@@ -199,7 +199,7 @@
   revision = "8d44bfdf1030639ae7130922c95df12d6d4da3b6"
 
 [[projects]]
-  name = "github.com/pingcap/kvproto"
+  name = "github.com/wentaoxu/kvproto"
```
Comment (zhangjinpeng1987) [inline Gopkg.lock]: Please update kvproto.

Comment (wentaoxu) [reply Gopkg.lock]: ok

File: store/tikv/gcworker/gc_worker.go [hunk 4]
```diff
@@ -467,11 +477,17 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
...
-		if len(key) == 0 {
-			break
+		if len(locks) < gcScanLockLimit {
+			regions++
+			key = loc.EndKey
+			if len(key) == 0 {
+				break
+			}
+		} else {
+			// if len(locks) is '0', we should get into the branch above, not here.
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Redundant comments, the code has talked.

File: store/tikv/gcworker/gc_worker.go [hunk 5]
```diff
@@ -425,6 +431,10 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
 		default:
 		}
 
+		// we'd better keep 'req.ScanLock.StartKey' the same as 'key' to avoid
+		// the error 'key is not in region' when the region split for several pieces, or
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Slit and merge cannot cause `key is not in region` error, but region error? Please confirm.

**Comment by zhangjinpeng1987** [comment]
Please add some metrics for resolve lock, for example how many region errors have encountered, weather resolve all lock successfully, and else errors have encountered.

**Comment by zhangjinpeng1987** [comment]
Please add tests to cover the case which one region need mutiple requests.

File: store/tikv/gcworker/gc_worker.go (hunk 5 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: `region error` will not too.
In the code below
```
if regionErr != nil {
   ...
   continue
}
```
then we will retry and get the  right region.

File: store/tikv/gcworker/gc_worker.go (hunk 4 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok

**Comment by wentaoxu** [comment]
@zhangjinpeng1987 it is hard to add test unless moving `resolveLock` out of the package `gc_worker`, because we can't call `prewrite` here.

**Comment by wentaoxu** [comment]
@zhangjinpeng1987 
1. #5694 this pr add metrics for gc failure.
2. Now it will print log and error if gc failed.

I'm not sure whether it is a good idea to stat `region error` here. may be we should add it on some more common place?

File: store/tikv/gcworker/gc_worker.go (hunk 5 above)
Comment (zhangjinpeng1987) [reply store/tikv/gcworker/gc_worker.go]: I mean you should fix this comment.

File: store/tikv/gcworker/gc_worker.go [hunk 6]
```diff
@@ -405,10 +406,15 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
 
 func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, identifier string) error {
 	gcWorkerCounter.WithLabelValues("resolve_locks").Inc()
+
+	// for scan lock request, we must return all locks even if they are generated
+	// by the same transaction. because gc worker need to make sure all locks to be
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: All locks have been cleaned.

File: store/tikv/gcworker/gc_worker.go (hunk 5 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok, I will add the situation to the comments.

File: store/tikv/gcworker/gc_worker.go (hunk 6 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok

**Comment by wentaoxu** [comment]
PTAL @zhangjinpeng1987

**Comment by disksing** [comment]
LGTM.

File: store/tikv/gcworker/gc_worker.go [hunk 7]
```diff
@@ -425,6 +431,12 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
 		default:
 		}
 
+		// We'd better keep 'req.ScanLock.StartKey' the same as 'key' to avoid
+		// the error 'key is not in region' when the region split for several pieces, or
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Region splitting and merging will cause region error, not `key is not in region`. Please fix this misunderstanding comment.

Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok

**Comment by wentaoxu** [comment]
PTAL @AndreMouche @zhangjinpeng1987

File: store/tikv/gcworker/gc_worker.go [hunk 8]
```diff
@@ -470,11 +477,16 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
...
-		if len(key) == 0 {
-			break
+		if len(locks) < gcScanLockLimit {
+			regions++
+			key = loc.EndKey
+			if len(key) == 0 {
+				break
+			}
+		} else {
+			key = locks[len(locks)-1].Key
```
Comment (AndreMouche) [inline store/tikv/gcworker/gc_worker.go]: Should we change the key in this situtation? It seems we should still use the  current_region's start_key here?

Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: Yes, we should, if we don't set it here, and set the `start key` of `ScanLock` req to the key of end lock(we must), when the region split, we may find the `start key` is larger than the end key of the region, and we will get the error `key not in region`, then the gc job will exit this time.

**Pull request merged**